sympy/parsing/mathematica.py

```diff
@@ -42,6 +46,10 @@ def parse(s):
         (r"\A(-? *[\d\.]+)([a-zA-Z].*)\Z",  # Implied multiplication - 2a
         lambda m: parse(m.group(1)) + "*" + parse(m.group(2))),
 
+        # Multiplication between any object and a function
+        (r"\A(.+)([*])(\w+\[[^\]]+[^\[]*\])\Z",
```

Comment:
What about `a Sin[x]`? If I recall correctly, there is some way of putting alternative matching conditions in the regex.